tests/drivers/clock_control/clock_control_api/src/test_clock_control.c

```diff
@@ -0,0 +1,310 @@
+
+typedef bool (*test_capability_check_t)(const char *dev_name);
+
+static void setup_instance(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+	int err;
+
+	k_busy_wait(1000);
+	do {
+		err = clock_control_off(dev, 0);
+	} while (err == 0);
+	LOG_INF("setup done");
+}
+
+static void tear_down_instance(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+
+	clock_control_on(dev, NULL);
+}
+
+static void test_all_instances(test_func_t func,
+				test_capability_check_t capability_check)
+{
+	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
+		if ((capability_check == NULL) ||
+			capability_check(devices[i].name)) {
+			setup_instance(devices[i].name);
+			func(devices[i].name, devices[i].startup_us);
+			tear_down_instance(devices[i].name);
+			/* Allow logs to be printed. */
+			k_sleep(100);
+		}
+	}
+}
+
+/*
+ * Basic test for checking correctness of getting clock status.
+ */
+static void test_on_off_status_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int err;
+
+	zassert_true(dev != NULL, "%s: Unknown device", dev_name);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+
+	err = clock_control_on(dev, NULL);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true((status == CLOCK_CONTROL_STATUS_STARTING) ||
+			(status == CLOCK_CONTROL_STATUS_ON),
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	k_busy_wait(startup_us*1.5);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_ON, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	err = clock_control_off(dev, 0);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+}
+
+static void test_on_off_status(void)
+{
+	test_all_instances(test_on_off_status_instance, NULL);
+}
+
+/*
+ * Test validates that if number of enabling requests matches disabling requests
+ * then clock is disabled.
+ */
+static void test_multiple_users_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int users = 5;
+	int err;
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	for (int i = 0; i < users; i++) {
+		err = clock_control_on(dev, NULL);
+		zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+	}
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true(status != CLOCK_CONTROL_STATUS_OFF,
```

Comment:
Could we specify here what we expect not what we not expect?

Reply:
at this point clock might be `STARTING` or `ON`. It's hard to determine. If clock is starting fast then it may be on already.

Reply:
So we should have `status == STARTING || status == ON`. This will be easier to read.

Reply:
done